### PR TITLE
[WIP] Adding AnnotatedServiceBuilder to add service level settings to annotated service

### PR DESCRIPTION
**Pull request by sivaalli**
Motivation:

It is good to provide an ability to user to configure per service level settings for annotated service.

Closes: https://github.com/line/armeria/issues/2023

Modifications:

1. Pulled all methods that `AbstractServiceBindingBuilder`(minus AbstractBindingBuilder)  has into an interface called `ServiceBuilder`
2. `AbstractServiceBindingBuilder` implements this(ServiceBuilder) interface and also extends `AbstractBindingBuilder`
3. `AbstractServiceBuilder` provides default implementation of `ServiceBuilder`

But there is still a code duplication in `AbstractServiceBindingBuilder` and `AbstractServiceBuilder`.

Result:
AnnotatedService can now be configured to have service specific properties.

This PR is not a **final** . This is just to get an initial review before making test and documentation changes.

### CODE REVIEW

**Comment by sivaalli** [comment]
Probably have to fix commit history. Will be fixing it in mean while.

File: core/src/main/java/com/linecorp/armeria/server/ServiceBuilder.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    ServiceBuilder requestContentPreviewerFactory(ContentPreviewerFactory factory);
+
+    /**
+     * Sets the {@link ContentPreviewerFactory} for an HTTP response of a {@link Service}.
+     * If not set, {@link VirtualHost#responseContentPreviewerFactory()}
+     * is used.
+     */
+    ServiceBuilder responseContentPreviewerFactory(ContentPreviewerFactory factory);
+    /**
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/ServiceBuilder.java]: Add a blank line?

File: core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java [hunk 2]
```diff
@@ -913,127 +913,15 @@ public final ServerBuilder service(
...
-
-            // If there is a decorator, we should add one more decorator which handles an exception
-            // raised from decorators.
-            if (s != e.service()) {
-                s = e.service().exceptionHandlingDecorator().apply(s);
-            }
-            service(e.route(), s);
-        });
-        return this;
-    }
```
[the remark this answers is not in the document]
Comment (sivaalli) [reply core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java]: Sure, your comment totally makes sense. Will add these back

**Comment by sivaalli** [comment]
> > But there is still a code duplication in AbstractServiceBindingBuilder and AbstractServiceBuilder.
> 
> In that case, I think we can use delegation:
> 
> ```java
> public abstract class AbstractServiceBindingBuilder extends AbstractBindingBuilder implements ServiceBuilder {
>     private final DefaultServiceBuilder serviceBuilder = new DefaultServiceBuilder();
>     ...
>     public AbstractServiceBindingBuilder requestTimeout(Duration requestTimeout) {
>         serviceBuilder.requestTimeout(requestTimeout);
>         return this;
>     }
>     ...
> }
> ```
> 
> WDYT?

Sure.

**Comment by sivaalli** [comment]
Will be reopening with clean commit history and implementing all suggestions.

**Comment by minwoox** [comment]
Hello @sivaalli! How's everything going? Just let us know if there's any trouble while you implementing this.

Meanwhile, I found out that @heowc's work on #1752 has a high chance of a large conflict with this PR. Because #1752 and this PR have to move the annotations parsing logic into `ServerBuilder.build()` I'm a little bit worried. 😉

**Comment by sivaalli** [comment]
Hi @minwoox , Everything is going good. I have made some progress. Currently adding tests. Expecting to be done by tomorrow.

**Comment by minwoox** [comment]
Glad to hear that. 😄 Then, I think @heowc needs to wait and get work after your PR is merged.

**Pull request closed without merging**